tfx/examples/cifar10_pipeline/cifar10_utils.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2019 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Python source file include CIFAR-10 pipeline functions and necessary utils.
+
+For a TFX pipeline to successfully run, a preprocessing_fn and a
+_build_estimator function needs to be provided.  This file contains both.
+"""
+
+import tensorflow as tf
+import tensorflow_model_analysis as tfma
+import tensorflow_transform as tft
+from tensorflow_transform.tf_metadata import schema_utils
+
+_IMAGE_KEY = 'image_raw'
+_LABEL_KEY = 'label'
+
+
+def _transformed_name(key):
+    if key == 'image_raw':
+        return key + '_xf_input'
+    return key + '_xf'
+
+
+def _gzip_reader_fn(filenames):
+    return tf.data.TFRecordDataset(
+        filenames,
+        compression_type='GZIP')
+
+
+def _get_raw_feature_spec(schema):
+    return schema_utils.schema_as_feature_spec(schema).feature_spec
+
+
+def _fill_in_missing(x):
+    default_value = '' if x.dtype == tf.string else 0
+    return tf.squeeze(
+        tf.sparse.to_dense(
```

Comment:
Now that the schema contains shape info, we are directly reading DenseFeature instead of SparseFeature, so this is not necessary I think.